pkg/util/fast_int_map.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2018 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package util
+
+import "golang.org/x/tools/container/intsets"
+
+// These constants determine the "small" representation: we pack <numVals>
+// values of <numBits> bits into <numWords> 64-bit words. Each value is 0 if the
+// corresponding key is not set, otherwise it is the value+1.
+//
+// It's desirable for efficiency that numBits, numValsPerWord are powers of two.
+//
+// The current settings support a map from keys in [0, 31] to values in [0, 14].
+// Note that one value is reserved to indicate an unmapped element.
+const (
+	numWords       = 2
+	numBits        = 4
+	numValsPerWord = 64 / numBits              // 16
+	numVals        = numWords * numValsPerWord // 32
+	mask           = (1 << numBits) - 1
+	maxValue       = mask - 1
+)
+
+// FastIntMap is a replacement for map[int]int which is more efficient
+// when the values are small. It can be passed by value.
+type FastIntMap struct {
+	small [numWords]uint64
+	large map[int]int
+}
+
+// returns -1 if the value is unmapped
+func (m *FastIntMap) getSmallVal(idx uint32) int32 {
```

Comment:
NIT: Not sure if you guys have a convention, but I always prefer to see methods roughly sorted by importance, from top to bottom (i.e. most important on top). In this case, I'd expect to see the exported methods first, since they're what I'd initially focus on.

Reply:
We don't have a convention, but that makes sense, moved them.